peerstore/peerstore.go

```diff
@@ -122,6 +121,74 @@ type AddrBook interface {
 	PeersWithAddrs() peer.IDSlice
 }
 
+// CertifiedAddrBook manages "self-certified" addresses for remote peers.
+// Self-certified addresses are contained in peer.PeerRecords
+// which are wrapped in a record.Envelope and signed by the peer
+// to whom they belong.
+//
+// Certified addresses (CA) are generally more secure than uncertified
+// addresses (UA). Consequently, CAs beat and displace UAs. When the
+// peerstore learns CAs for a peer, it will reject UAs for the same peer
+// (as long as the former haven't expired).
+// Furthermore, peer records act like sequenced snapshots of CAs. Therefore,
+// processing a peer record that's newer than the last one seen overwrites
+// all addresses with the incoming ones.
+//
+// This interface is most useful when combined with AddrBook.
+// To test whether a given AddrBook / Peerstore implementation supports
+// certified addresses, callers should use the GetCertifiedAddrBook helper or
+// type-assert on the CertifiedAddrBook interface:
+//
+//     if cab, ok := aPeerstore.(CertifiedAddrBook); ok {
+//         cab.ProcessPeerRecord(signedPeerRecord, aTTL)
+//     }
+//
+type CertifiedAddrBook interface {
+	// ProcessPeerRecord adds addresses from a signed peer.PeerRecord (contained in
+	// a routing.Envelope), which will expire after the given TTL.
+	//
+	// Signed records added via this method will be stored without
+	// alteration as long as the address TTLs remain valid. The Envelopes
+	// containing the PeerRecords can be retrieved by calling GetPeerRecord(peerID).
+	//
+	// If the signed PeerRecord belongs to a peer that already has certified
+	// addresses in the CertifiedAddrBook, the new addresses will replace the
+	// older ones, iff the new record has a higher sequence number than the
+	// existing record. Attempting to add a peer record with a
+	// sequence number that's <= an existing record for the same peer will not
+	// result in an error, but the record will be ignored.
+	//
+	// If the CertifiedAddrBook is also an AddrBook (which is most likely the case),
+	// adding certified addresses for a peer will *replace* any
+	// existing non-certified addresses for that peer, and only the certified
+	// addresses will be returned from AddrBook.Addrs thereafter.
+	//
+	// Likewise, once certified addresses have been added for a given peer,
+	// any non-certified addresses added via AddrBook.AddAddrs or
+	// AddrBook.SetAddrs will be ignored. AddrBook.SetAddrs may still be used
+	// to update the TTL of certified addresses that have previously been
+	// added via ProcessPeerRecord.
+	ProcessPeerRecord(s *record.Envelope, ttl time.Duration) error
```

Comment:
`ConsumePeerRecord`? `Process` sounds kind of generic (I know I had proposed that name, but I'm having second thoughts)

Reply:
Can we make this method return (accepted bool, err error), where accepted indicates if this record was accepted?